File: src/001_domain_model_pattern/order.java
public class Order {
    private OrderState state;
    private ShippingInfo shippingInfo;

    public void changeShippingInfo(ShippingInfo newShippingInfo) {
        if (!state.isShippingChangeable()) {
            throw new IllegalStateException("Can't change shipping in " + state)
        }

        this.shippingInfo = newShippingInfo;
    }
}

public enum OrderState {
    PAYMENT_WAITING {
        public boolean isShippingChangeable() {
            return true;
        }
    },
    PREPARING {
        public boolean isShippingChangeable() {
            return true
        }
    },
    SHIPPED, DELIVERYING, DELIVERY_COMPLETED;
    public boolean isShippingChangeable() {
        return false;
    }
}

File: src/1_domain_model_pattern/order.ts
enum OrderState {
  PAYMENT_WAITING,
  PREPARING,
  SHIPPED,
  DELIVERING,
  DELIVERY_COMPLETED,
}

class IllegalStateException extends Error {}

class ShippingInfo {}

class Order {
  private state: OrderState;
  private shippingInfo: ShippingInfo;

  public changeShippingInfo(newShippingInfo: ShippingInfo): void {
    if (!this.isShippingChangeable())
      throw new IllegalStateException("Can't change shipping in " + this.state);

    this.shippingInfo = newShippingInfo;
  }

  private isShippingChangeable(): boolean {
    return (
      this.state == OrderState.PAYMENT_WAITING ||
      this.state == OrderState.PREPARING
    );
  }
}
